### PR TITLE
Improve the density test of known bad items

### PR DESCRIPTION
**Pull request by mqrause**
#### Summary
None

#### Purpose of change

This test is causing some confusion. This attempts to reduce the confusion by improving the fail message and forcing contributors to prune the known_bad list when they fix items.

#### Describe the solution

- remove !mayfail from the test
- jsonify the list of bad items
- remove fixed items from the list
- improve the info message when the test fails to hopefully be descriptive enough

#### Describe alternatives you've considered

With the bad list jsonified, a further improvement could be to remove the test altogether and make all this a sanity check on item definitions. That would give people instant feedback when they test their changes.

#### Testing

Ran the `item_material_density_blacklist_is_pruned` after removing !mayfail, then removed the fixed items and ran it again.

### CODE REVIEW

File: tests/item_test.cpp [hunk 1]
```diff
@@ -1930,7 +1923,8 @@ TEST_CASE( "item_material_density_blacklist_is_pruned", "[item][!mayfail]" )
         }
         const std::map<material_id, int> &mats = target.made_of();
         if( !mats.empty() ) {
-            INFO( bad.str() );
+            INFO( string_format( "%s had its density fixed, remove it from the known_bad list in tests\\item_test.cpp",
```
Comment (anothersimulacrum) [inline tests/item_test.cpp:1926]: GitHub displays unix-style paths.
```suggestion
            INFO( string_format( "%s had its density fixed, remove it from the known_bad list in tests/item_test.cpp",
```

**Review by anothersimulacrum — COMMENTED** [review]
I don't think this should cause people's PRs to fail, because I'm worried that for some contributors editing test files could be problematic (e.g. their editor causes issues, they can't actually compile and test it, etc).

**Comment by mqrause** [comment]
> I don't think this should cause people's PRs to fail, because I'm worried that for some contributors editing test files could be problematic (e.g. their editor causes issues, they can't actually compile and test it, etc).

The main reason I think this should cause failures is because as it is it leads to errors showing up on unrelated PRs, which causes confusion and possibly masks actual errors they need to fix. If making people edit test files is a no-no, I'll look into jsonifying the list.

File: src/itype.cpp [hunk 2]
```diff
@@ -376,3 +376,13 @@ std::tuple<encumbrance_modifier_type, int> armor_portion_data::convert_descripto
 }
 
 std::map<itype_id, std::set<itype_id>> islot_magazine::compatible_guns;
+
+std::set<itype_id> known_bad_density::known_bad;
+
+void known_bad_density::load( const JsonObject &jo )
+{
+    jo.allow_omitted_members();
```
Comment (anothersimulacrum) [inline src/itype.cpp:384]: I don't think you need this.
```suggestion
```

**Review by anothersimulacrum — APPROVED** [review]

File: src/itype.cpp (hunk 2 above)
Comment (mqrause) [reply src/itype.cpp:384]: Yeah, probably. I just copied it from the charge removal/migration blacklists.

**Pull request merged**